File: android/src/main/kotlin/io/storecamera/storecamera_photo/storecamera_plugin_camera/data/CameraSetting.kt
package io.storecamera.storecamera_photo.storecamera_plugin_camera.data

data class CameraSetting(
        val currentPosition: CameraPosition,
        val hasFlashUnit: Boolean,
        val minZoom: Double,
        val maxZoom: Double
) : PluginToMap {

    override fun pluginToMap(): Map<String, *> {
        val hasFlashUnit: Boolean = if(currentPosition == CameraPosition.BACK) this.hasFlashUnit else false

        return hashMapOf(
                "currentPosition" to currentPosition.value,
                "isTorchAvailable" to hasFlashUnit,
                "isFlashAvailable" to hasFlashUnit,
                "minZoom" to minZoom,
                "maxZoom" to maxZoom,
                "isExposureAvailable" to false,
                "minExposure" to 0.0,
                "maxExposure" to 0.0,
                "currentExposure" to 0.0,
        )
    }
}

enum class CameraPosition(val value: String) {
    BACK("BACK"),
    FRONT("FRONT");

    companion object {
        fun from(value: String?): CameraPosition? {
            for(_value in values()) {
                if(_value.value == value) {
                    return _value
                }
            }
            return null
        }
    }
}

enum class CameraRatio(val value: String) {
    RATIO_1_1("1:1"),
    RATIO_4_5("4:5"),
    RATIO_3_4("3:4"),
    RATIO_9_16("9:16");

    companion object {
        fun from(value: String?): CameraRatio? {
            for(_value in values()) {
                if(_value.value == value) {
                    return _value
                }
            }
            return null
        }
    }
}

enum class CameraResolution(val value: String) {
    FULL("FULL"),
    STANDARD("STANDARD"),
    LOW("LOW");

    companion object {
        fun from(value: String?): CameraResolution? {
            for(_value in values()) {
                if(_value.value == value) {
                    return _value
                }
            }
            return null
        }
    }
}

enum class CameraFlash(val value: String) {
    OFF("OFF"),
    ON("ON"),
    AUTO("AUTO");

    companion object {
        fun from(value: String?): CameraFlash? {
            for(_value in values()) {
                if(_value.value == value) {
                    return _value
                }
            }
            return null
        }
    }
}

enum class CameraTorch(val value: String) {
    OFF("OFF"),
    ON("ON");

    companion object {
        fun from(value: String?): CameraTorch? {
            for(_value in values()) {
                if(_value.value == value) {
                    return _value
                }
            }
            return null
        }
    }
}

enum class CameraMotion(val value: String) {
    OFF("OFF"),
    ON("ON");

    companion object {
        fun from(value: String?): CameraMotion? {
            for(_value in values()) {
                if(_value.value == value) {
                    return _value
                }
            }
            return null
        }
    }
}

File: ios/Classes/camera/data/CameraSetting.swift
//
//  CameraSetting.swift
//  store_camera_plugin_camera
//
//  Created by 김경환 on 2020/10/28.
//

import AVFoundation

struct CameraSetting: CameraPluginToMap {
    
    let currentPosition: AVCaptureDevice.Position

    let isTorchAvailable: Bool
    let isFlashAvailable: Bool
    let minZoom: Float
    let maxZoom: Float

    let minExposure: Float
    let maxExposure: Float
    let currentExposure: Float

    func pluginToMap() -> [String : Any] {
        let isTorchAvailable: Bool
        let isFlashAvailable: Bool
        if currentPosition == AVCaptureDevice.Position.back {
            isTorchAvailable = self.isTorchAvailable
            isFlashAvailable = self.isFlashAvailable
        } else {
            isTorchAvailable = false
            isFlashAvailable = false
        }
        
        return [
            "currentPosition": positionToString(self.currentPosition) ?? "",
            "isTorchAvailable": isTorchAvailable,
            "isFlashAvailable": isFlashAvailable,
            "minZoom": minZoom,
            "maxZoom": maxZoom,
            "isExposureAvailable": true,
            "minExposure": minExposure,
            "maxExposure": maxExposure,
            "currentExposure": currentExposure,
        ]
    }
    
    public init(_ device: AVCaptureDevice) {
        self.currentPosition = device.position
        self.isTorchAvailable = device.isTorchAvailable
        self.isFlashAvailable = device.isFlashAvailable
        self.minZoom = Float(device.minAvailableVideoZoomFactor)
        self.maxZoom = Float(device.maxAvailableVideoZoomFactor)
        self.minExposure = device.minExposureTargetBias
        self.maxExposure = device.maxExposureTargetBias
        self.currentExposure = device.exposureTargetBias
    }
}

func positionToString(_ position: AVCaptureDevice.Position) -> String? {
    switch position {
    case .unspecified:
        return nil
    case .back:
        return "BACK"
    case .front:
        return "FRONT"
    @unknown default:
        return nil
    }
}

enum CameraRatio: String {
    case RATIO_1_1 = "1:1"
    case RATIO_4_5 = "4:5"
    case RATIO_3_4 = "3:4"
    case RATIO_9_16 = "9:16"
}

enum CameraResolution: String {
    case FULL = "FULL"
    case STANDARD = "STANDARD"
    case LOW = "LOW"
}

enum CameraFlash: String {
    case OFF = "OFF"
    case ON = "ON"
    case AUTO = "AUTO"
    
    func deviceMode() -> AVCaptureDevice.FlashMode {
        switch self {
        case .OFF:
            return AVCaptureDevice.FlashMode.off
        case .ON:
            return AVCaptureDevice.FlashMode.on
        case .AUTO:
            return AVCaptureDevice.FlashMode.auto
        }
    }
}

enum CameraTorch: String {
    case OFF = "OFF"
    case ON = "ON"
    
    func deviceMode() -> AVCaptureDevice.TorchMode {
        switch self {
        case .OFF:
            return AVCaptureDevice.TorchMode.off
        case .ON:
            return AVCaptureDevice.TorchMode.on
        }
    }
}

enum CameraMotion: String {
    case OFF = "OFF"
    case ON = "ON"
}
